### PR TITLE
Leticia Tran (Ampers)

### PR DESCRIPTION
**Pull request by LetiTran**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I started with a single method for requesting new trips, but soon realized I should probably have it be sort of a "calling"  method that would call of the helper methods that were necessary for requesting a new trip, such as checking that both the passenger and the driver with the right requirements are available, creating a new trip, populating all of the trips list with this new trip and changing the driver's status for unavailable. This made it possible to follow the single responsibility requirement for a well designed method.|
Describe any examples of composition that you encountered in this project, if any | This whole project has it's Classes tied together, where the trips have passengers and drivers, the drivers have passengers and trips, and the passengers have trips with drivers. |
Describe a _nominal test_ that you wrote for this assignment. | The test that asserts that my method for requesting a new trip is creating a new instance of trip that has passengers and drivers too. |
Describe an _edge case test_ that you wrote for this assignment. | The test that returns an StandardError if there are no available drivers for the new trip requests. |
Describe a concept that you gained more clarity on as you worked on this assignment. | I was able to use more enumerables and it helped me understand how to use them a little more. |

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -13,9 +14,20 @@ def initialize(input)
...
-        raise ArgumentError.new("Invalid rating #{@rating}")
+      if @rating != nil
+        raise ArgumentError.new("Invalid rating #{@rating}")  if @rating > 5 || @rating < 1
       end
+
+      raise ArgumentError.new("End time cannot be before the start time.") if @start_time > @end_time if @end_time != nil
+    end
+
+    def trip_duration
+      return @end_time == nil ? nil : (@end_time.to_f - @start_time.to_f)
```
Comment (CheezItMan) [inline lib/trip.rb:25]: 👍

File: lib/passenger.rb [hunk 2]
```diff
@@ -18,7 +16,23 @@ def get_drivers
...
     def add_trip(trip)
+      # Add trip to passangers list:
+      raise ArgumentError.new("Can only add trip instance to trip collection") if trip.class != Trip
       @trips << trip
     end
+
+    def total_spent
+      # Passanger's total $ spent so far:
+      total_money = 0
+      @trips.each {|trip| trip.end_time == nil ? total_money += 0 : total_money += trip.cost}
```
Comment (CheezItMan) [inline lib/passenger.rb:27]: An inject or sum would work a bit neater.

File: specs/trip_spec.rb [hunk 3]
```diff
@@ -38,5 +38,62 @@
...
+        end_time: end_time,
+        cost: 23.45,
+        rating: 3
+      }
+
+      proc {RideShare::Trip.new(@trip_data)}.must_raise ArgumentError
+    end
+  end
+
+  describe '#trip_duration' do
```
Comment (CheezItMan) [inline specs/trip_spec.rb:61]: I would also test what happens if there are no trips.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -77,4 +82,65 @@
       driver.average_rating.must_equal 0
     end
   end
+
+  describe '#total_rev' do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:86]: Also testing when there are no trips would be good.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -77,4 +82,65 @@
...
+
+      @driver.add_trip(trip)
+
+      total_hours = (end_time.to_i - start_time.to_i)/3600
+      total_rev = (10 - 1.65) * 0.8
+      average_rev = total_rev / total_hours
+      @driver.average_revenue.must_equal average_rev
+    end
+
+    it 'Ignores the trips that are still in progress' do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:133]: 👍

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -88,5 +88,244 @@
...
+      find_new_trip_in_passanger.must_equal @new_trip
+    end
+
+    it 'Selects an AVAILABLE driver' do
+
+      initial_drivers_list = @trip_disp.load_drivers
+
+      # Find the status for this driver in the initial list from file:
+      index = @new_trip.driver.id - 1
+      initial_status = initial_drivers_list[index].status
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:156]: Be a little careful about things like this.  This depends on the fact that the CSV file is in order and has no missing ids.  If someone deleted a driver, this wouldn't work.

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -88,5 +88,244 @@
...
+
+  describe 'wave3_request_trip' do
+    before do
+      @trip_disp = RideShare::TripDispatcher.new()
+      # Request new trip:
+      @new_trip = @trip_disp.wave3_request_trip(1)
+    end
+
+    it 'Creates a new instance of trips' do
+      @new_trip.must_be_instance_of RideShare::Trip
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:178]: Notice that a lot of these are a re-do of the above.  You could create methods that each test calls.

File: lib/trip_dispatcher.rb [hunk 8]
```diff
@@ -86,16 +91,160 @@ def load_trips
...
+    def request_trip(passenger_id)
+      # Initiate a new trip:
+      passenger = find_passenger(passenger_id)
+
+      raise ArgumentError.new("Passanger with id #{passenger_id} does not exist.") if passenger == nil
+
+      begin
+        select_driver_available
+      rescue SystemCallError => exception
+        puts "There are no available drivers #{exception.message}"
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:112]: puts?  No, you should either return something (maybe `nil`) or raise an error.

File: lib/trip_dispatcher.rb [hunk 9]
```diff
@@ -86,16 +91,160 @@ def load_trips
...
+        last = driver.trips.max_by { |trips| trips.end_time.to_i }
+        lastest_trips_of_all_drivers << [driver, last]
+      end
+      return lastest_trips_of_all_drivers
+    end
+
+    def select_the_right_driver
+
+      # Return a driver with no trips if it exists:
+      select_driver_with_no_trips
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:237]: You're not doing anything with this method call.

File: lib/trip_dispatcher.rb [hunk 10]
```diff
@@ -86,16 +91,160 @@ def load_trips
...
+    end
+
+    #________________________Private methods for WAVE 3:
+    def select_driver_with_earliest_end_of_trip(only_available)
+      earliest_end_time = Time.now
+      right_driver = nil
+
+      only_available.each do |sub_array|
+
+        return right_driver = sub_array[0] if sub_array[1] == nil
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:202]: I'm not sure this is needed.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and good commit messages
Answer comprehension questions | Check
**Wave 1** |
`Trip` has a helper method to calculate duration | Check, nice ternary
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue | Check
Complete tests | Yes, but see my notes, you should have a test for when there are no trips as well.
**Wave 2** |
`TripDispatcher` has a new method to create trips | Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | Well done!
Complete tests | Check, 
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | Yes, although you are ignoring drivers without trips I believe.  
Appropriate use of Ruby's `Time` | Check
Appropriate helper methods were made to help with complex logic | Lots of helper methods
Complete tests | Good tets
**Overall** | Really nice work, you hit all the learning goals.  Good work with Wave 3.  I left some minor comments, but you did very well here.
